### PR TITLE
Fix breaking change in fixed AddressOf expressions

### LINKED ISSUE #27945: Investigate PR 27424 Breaking change around unfixed expressions
PR #27424 caused a build break for the following scenario:

```csharp
using System;
public unsafe sealed class PinnedStreamReader
{
    public unsafe byte* GetPointer(int position)
    {
        return (&bufferStart[position]);
    }
 	
    byte* bufferStart;
}
```

It now produced: `error CS0212: You can only take the address of an unfixed expression inside of a fixed statement initializer`

PR #27944 reverted it in master. We should investigate this and bring back the fix, without breaking additional scenarios.

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #27945

* First commit brings back the old PR.
* Second one fixes the breaking change and adds tests.
* Third one just renames the API along with the tests. Reason is below:

`IsNonMoveableVariable()` is confusing AND incorrect. The name indicates that such variables/expressions cannot be moved during runtime, but it does return true for things that can be moved. One example is the breaking change. A pointer field of a type can be moved, but the compiler/language rules allow it.

Renaming to `ExpressionRequiresFixing()` indicates the intention of the compiler in a more straight-forward way. It will always return whether some expression needs to be fixed (by a `fixed` statement) or not. Tests were updated accordingly.

### CODE REVIEW

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler for review.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 1]
```diff
@@ -2225,19 +2224,32 @@ internal bool IsNonMoveableVariable(BoundExpression expr, out Symbol accessedLoc
...
                     case BoundKind.PointerIndirectionOperator: //Covers ->, since the receiver will be one of these.
-                    case BoundKind.PointerElementAccess:
                     case BoundKind.ConvertedStackAllocExpression:
                         {
-                            return true;
+                            return false;
+                        }
+                    case BoundKind.PointerElementAccess:
+                        {
+                            BoundExpression underlyingExpr = ((BoundPointerElementAccess)expr).Expression;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: Probably deserves a comment.

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler any other reviewers?

**Comment by jaredpar** [comment]
Have we verified that this version of the fix can successfully build the DevDiv repository? This change is going in pretty late and it's caused a regression before. Want to know what our confidence level is here.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 2]
```diff
@@ -2233,6 +2233,8 @@ internal bool ExpressionRequiresFixing(BoundExpression expr, out Symbol accessed
                         }
                     case BoundKind.PointerElementAccess:
                         {
+                            // When using pointers in unsafe code, it is generally left for the user to make sure memory is fixed/pinned correctly.
+                            // Unless for fixed fields, we want to produce appropriate diagnostics if field is not fixed before access.
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: Typo: Unless -> Except

**Comment by OmarTawfik** [comment]
@jaredpar I queued a validation build and I didn't see any breaks.

**Comment by jcouv** [comment]
@OmarTawfik This PR seems ready to merge

**Pull request merged**